File: src/main/java/rails/game/action/TakeLoans.java
package rails.game.action;

import java.io.IOException;
import java.io.ObjectInputStream;

import com.google.common.base.Objects;

import net.sf.rails.game.PublicCompany;
import net.sf.rails.util.RailsObjects;

/**
 * Rails 2.0: Updated equals and toString methods
 */
public class TakeLoans extends PossibleORAction {

    // Initial attributes
    // TODO: This is a duplication of the field in PossibleORAction
    // Is there a reason for that? (potentially that it could be used outside of ORs)
    transient private PublicCompany company;
    private String companyName;
    private int maxNumber;
    private int price;

    // User-assigned attributes
    private int numberTaken = 0;

    public static final long serialVersionUID = 1L;

    /**
     *
     */
    public TakeLoans(PublicCompany company, int maxNumber,
            int price) {

        this.company = company;
        this.companyName = company.getId();
        this.maxNumber = maxNumber;
        this.price = price;
    }

    /**
     * @return Returns the minimumPrice.
     */
    public int getMaxNumber() {
        return maxNumber;
    }

    /**
     * @return Returns the company.
     */
    @Override
    public PublicCompany getCompany() {
        return company;
    }

    public int getPrice() {
        return price;
    }

    public void setNumberTaken(int numberTaken) {
        this.numberTaken = numberTaken;
    }

    public int getNumberTaken() {
        return numberTaken;
    }

    @Override
    protected boolean equalsAs(PossibleAction pa, boolean asOption) {
        // identity always true
        if (pa == this) return true;
        //  super checks both class identity and super class attributes
        if (!super.equalsAs(pa, asOption)) return false; 

        // check asOption attributes
        TakeLoans action = (TakeLoans)pa; 
        boolean options = 
                Objects.equal(this.company, action.company)
                && Objects.equal(this.maxNumber, action.maxNumber)
                && Objects.equal(this.price, action.price)
        ;
        
        // finish if asOptions check
        if (asOption) return options;
        
        // check asAction attributes
        return options
                && Objects.equal(this.numberTaken, action.numberTaken)
        ;
    }

    @Override
    public String toString() {
        return super.toString() + 
                RailsObjects.stringHelper(this)
                    .addToString("company", company)
                    .addToString("maxNumber", maxNumber)
                    .addToString("price", price)
                    .addToStringOnlyActed("numberTaken", numberTaken)
                    .toString()
        ;
    }

    /** Deserialize */
    private void readObject(ObjectInputStream in) throws IOException,
            ClassNotFoundException {

        in.defaultReadObject();

        company =
                getCompanyManager().getPublicCompany(companyName);
    }

}

File: dotnet/RailsLib.Net/Rails/Game/Action/TakeLoans.cs
using GameLib.Net.Game;
using GameLib.Net.Util;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using System.Text;

namespace GameLib.Rails.Game.Action
{
    [JsonObject(MemberSerialization.Fields)]
    public class TakeLoans : PossibleORAction
    {
        // Initial attributes
        // TODO: This is a duplication of the field in PossibleORAction
        // Is there a reason for that? (potentially that it could be used outside of ORs)
        /*transient*/
        [JsonIgnore]
        new private PublicCompany company;
        new private string companyName;
        private int maxNumber;
        private int price;

        // User-assigned attributes
        private int numberTaken = 0;

        new public const long serialVersionUID = 1L;

        /**
         *
         */
        public TakeLoans(PublicCompany company, int maxNumber, int price)
        {
            this.company = company;
            this.companyName = company.Id;
            this.maxNumber = maxNumber;
            this.price = price;
        }

        /**
         * @return Returns the minimumPrice.
         */
        public int MaxNumber
        {
            get
            {
                return maxNumber;
            }
        }

        /**
         * @return Returns the company.
         */
        override public PublicCompany Company
        {
            get
            {
                return company;
            }
        }

        public int Price
        {
            get
            {
                return price;
            }
        }

        public int NumberTaken
        {
            get
            {
                return numberTaken;
            }
            set
            {
                numberTaken = value;
            }
        }

        override protected bool EqualsAs(PossibleAction pa, bool asOption)
        {
            // identity always true
            if (pa == this) return true;
            //  super checks both class identity and super class attributes
            if (!base.EqualsAs(pa, asOption)) return false;

            // check asOption attributes
            TakeLoans action = (TakeLoans)pa;
            bool options = company.Equals(action.company)
                    && (maxNumber == action.maxNumber)
                    && (price == action.price);

            // finish if asOptions check
            if (asOption) return options;

            // check asAction attributes
            return options && (numberTaken == action.numberTaken);
        }

        override public string ToString()
        {
            return base.ToString() +
                    RailsObjects.GetStringHelper(this)
                        .AddToString("company", company)
                        .AddToString("maxNumber", maxNumber)
                        .AddToString("price", price)
                        .AddToStringOnlyActed("numberTaken", numberTaken)
                        .ToString();
        }

        /** Deserialize */
        [OnDeserialized]
        new internal void OnDeserialized(StreamingContext context)
        {
            company = CompanyManager.GetPublicCompany(companyName);
        }
    }
}
